sw/host/tests/chip/sysrst_ctrl/sysrst_ctrl_outputs.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, Result};
+use clap::Parser;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+use std::fs;
+use std::path::PathBuf;
+use std::time::Duration;
+
+use object::{Object, ObjectSymbol};
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::io::uart::Uart;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::mem::MemWriteReq;
+use opentitanlib::test_utils::test_status::TestStatus;
+use opentitanlib::uart::console::UartConsole;
+use opentitanlib::{collection, execute_test};
+
+use sysrst_ctrl::{Config, setup_pins, set_pins, read_pins};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+
+    /// Path to the firmware's ELF file, for querying symbol addresses.
+    #[arg(value_name = "FIRMWARE_ELF")]
+    firmware_elf: PathBuf,
+}
+
+struct Params<'a> {
+    opts: &'a Opts,
+    transport: &'a TransportWrapper,
+    uart: &'a dyn Uart,
+    config: &'a Config,
+    test_phase_addr: u32,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+enum TestPhase {
+  // There is a Setup phase but this is the initial value on the device code so we do not need it.
+    Loopback = 1,
+    OverrideSetup = 2,
+    OverrideZeros = 3,
+    OverrideOnes = 4,
+    OverrideRelease = 5,
+    OverrideAndLoopback = 6,
+    Done = 7,
+}
+
+static CONFIG: Lazy<HashMap<&'static str, Config>> = Lazy::new(|| {
+    collection! {
+        "hyper310" => Config {
+            // key0_in, key1_in, key2_in, pwrb_in
+            output_pins: vec!["IOR10", "IOR11", "IOR12", "IOR5", ],
+            open_drain: vec![false, false, false, false],
+            // key0_out, key1_out, key2_out, pwrb_out, bat_dis_out, z3_wakeup_out, ec_rst, flash_wp
+            input_pins: vec!["IOR6", "IOR7", "IOB0", "IOB1", "IOB2", "IOB3", "IOR8", "IOR9"],
+            // ec_rst, flash_wp
+            pullup_pins: vec!["IOR8", "IOR9"],
+        },
+    }
+});
+
+const OUTPUT_ALL_SET: u8 = 0b11111111;
+const OUTPUT_NONE_SET: u8 = 0b00000000;
+const LOOPBACK_ALL_SET: u8 = 0b1111;
+const LOOPBACK_PARTIAL_SET: u8 = 0b1010;
+const LOOPBACK_PATTERN_LENGTH: usize = 16;
+
+fn set_loopback_pads(params: &Params, pad_values: u8) -> Result<()> {
+    set_pins(params.transport, params.config, pad_values)
+}
+
+fn read_loopback_pads(params: &Params) -> Result<u8> {
+    // Only capture the pins of interest.
+    Ok(read_pins(params.transport, params.config)? & LOOPBACK_ALL_SET)
+}
+
+fn read_output_pads(params: &Params) -> Result<u8> {
+    read_pins(params.transport, params.config)
+}
+
+fn check_loopback_single(params: &Params) -> Result<()> {
```

Comment:
Nit: I was slightly confused that the `check_...` functions also set the loopback, but not a huge issue. I'm not sure I have a constructive suggestion for a different name though, sorry!

Reply:
I followed closely the DV test. Would it help if I rename it to something like `check_loopback_works`? With a verb, it's clearer that it is not just reading?

Reply:
Hmm I think probably the current name is good enough, I won't bikeshed it too much